### ISSUE TITLE
TODO: fix numerical instability in sparse_softmax_cross_entropy_with_logits 

### ISSUE THREAD

**Issue by daquexian** [Oneflow-Inc/oneflow#3137]
sparse_softmax_cross_entropy_with_logits 里面包含一个 log softmax，oneflow 里会直接计算 SafeLog(softmax(x))，这样在一些情况下 softmax(x) 会低于 SafeLog 的阈值 1e-20，只能得到一个常数 SafeLog(1e-20)=-46.0517，但其他框架 tf、pytorch 计算 log_softmax 时会把 softmax(x) 的分子提到 log 外面，变成 
```
tmp = x - reduce_max(x, axis);
log_softmax(x, axis) = tmp - log(reduce_sum(exp(tmp), axis))
```
（可参考 scipy 的代码： https://github.com/scipy/scipy/blob/v1.5.0/scipy/special/_logsumexp.py#L217 ）
这样在 `axis` 方向上，`tmp` 一定有一个 0，并且其它值都小于 0，所以 `log(x)` 的输入 `reduce_sum(exp(tmp), axis)` 一定大于 1 而且小于 x.shape[axis]，是一个非常安全的范围，log 总能得到正确的结果。

一个可以复现问题的完整代码：

```python
import numpy as np
import oneflow as flow
import torch
import torch.nn.functional as F
import tensorflow as tf

func_config = flow.FunctionConfig()
func_config.default_data_type(flow.float)

@flow.global_function(func_config)
def FlowJob(labels=flow.FixedTensorDef((1,), dtype=flow.int64), logits=flow.FixedTensorDef((1, 2))):                                                              
  return flow.nn.sparse_softmax_cross_entropy_with_logits(labels, logits)
 
labels=np.array([0])
logits=np.array([[-30, 30]]).astype(np.float32)
tf_res = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels, logits=logits)
pytorch_res = F.cross_entropy(torch.tensor(logits), torch.tensor(labels))
flow_res = FlowJob(labels, logits).get().ndarray()
print(f'tf: {tf_res}')
print(f'pytorch: {pytorch_res}')
print(f'flow: {flow_res}') 
```

输出：

> tf: [60.]
> pytorch: 60.0
> flow: [46.0517]

🤣 我最近没有时间修，记个 TODO
